src/CodeGen_D3D12Compute_Dev.cpp

```diff
@@ -0,0 +1,1154 @@
+    src_stream.str("");
+    src_stream.clear();
+
+    // compiler control pragmas
+    src_stream
+        // warning X3078: 'i': loop control variable conflicts with a previous declaration in the outer scope; most recent declaration will be used
+        << "#pragma warning( disable : 3078 )" "\n"
+        // TODO(marcos): can we interchangeably replace ints by uints when we have modulo operations in the generated code?
+        // warning X3556 : integer modulus may be much slower, try using uints if possible
+        << "#pragma warning( disable : 3556 )" "\n"
+        << "\n";
+
+    // Write out the Halide math functions.
+    src_stream 
+             //<< "namespace {\n"   // HLSL does not support unnamed namespaces...
+               #if DEBUG_TYPES
+               << "#define  int8   int\n"
+               << "#define  int16  int\n"
+               << "#define  int32  int\n"
+               << "#define uint8  uint\n"
+               << "#define uint16 uint\n"
+               << "#define uint32 uint\n"
+               << "\n"
+               << "#define  bool_(x)   bool##x\n"
+               << "#define  int8_(x)   int##x\n"
+               << "#define  int16_(x)  int##x\n"
+               << "#define  int32_(x)  int##x\n"
+               << "#define uint8_(x)  uint##x\n"
+               << "#define uint16_(x) uint##x\n"
+               << "#define uint32_(x) uint##x\n"
+               << "\n"
+               << "#define asint32  asint\n"
+               << "#define asuint32 asuint\n"
+               << "\n"
+               #endif
+               << "float nan_f32()     { return( 1.#IND); } \n" // Quiet NaN with minimum fractional value.
+               << "float neg_inf_f32() { return(-1.#INF); } \n"
+               << "float inf_f32()     { return(+1.#INF); } \n"
+               << "#define float_from_bits asfloat \n"
+               << "#define sqrt_f32    sqrt   \n"
+               << "#define sin_f32     sin    \n"
+               << "#define cos_f32     cos    \n"
+               << "#define exp_f32     exp    \n"
+               << "#define log_f32     log    \n"
+               << "#define abs_f32     abs    \n"
+               << "#define floor_f32   floor  \n"
+               << "#define ceil_f32    ceil   \n"
+               << "#define round_f32   round  \n"
+               << "#define trunc_f32   trunc  \n"
+               << "#define pow_f32     pow    \n"
+               << "#define asin_f32    asin   \n"
+               << "#define acos_f32    acos   \n"
+               << "#define tan_f32     tan    \n"
+               << "#define atan_f32    atan   \n"
+               << "#define atan2_f32   atan2  \n"
+               << "#define sinh_f32    sinh   \n"
+               << "#define asinh_f32   asinh  \n"
+               << "#define cosh_f32    cosh   \n"
+               << "#define acosh_f32   acosh  \n"
+               << "#define tanh_f32    tanh   \n"
+             //<< "#define atanh_f32 atanh \n"
```

Comment:
is `atanh()` not available?  Maybe then define it as `#define atah_f32(x) (log_f32((1+x)/(1-x))/2)`